Fix for idea 233+, add ability specify file name prefix for an empty path

vlascik opened the pull request:
Fixes compatibility with Idea 233+ and most deprecations, and does some cleanup (typos, translations to english etc.).

Probably shouldn't contain the .jar files, but as I don't use Java really, I don't have time to experiment with some of the deprecations and builds.

I'm working on one more feature (file title regexps), so I will probably do more cleanup later, but I have no idea when, and some people might still find this plugin working better sooner than later.

If anyone needs a .jar, it's at https://github.com/vlascik/Tabdir/releases/download/release-1.7.0-mod-vl/tabdir-1.7.0-mod-vl.jar

crazyproger commented on META-INF/plugin.xml:
```diff
@@ -53,35 +53,32 @@
 
-        <projectService serviceImplementation="ru.crazycoder.plugins.tabdir.configuration.ProjectConfig"/>
+        <projectService serviceImplementation="ru.crazycoder.plugins.tabdir.TitleFormatter"/>
+        <projectConfigurable
```
Am I right that this block replaces ProjectConfigRegistrator?

vlascik replied:
yes, this seems to be  a new IDEA way to register configuration services

crazyproger commented on src/ru/crazycoder/plugins/tabdir/configuration/FolderConfiguration.java:
```diff
@@ -133,6 +137,11 @@ public void setRemoveDuplicates(boolean removeDuplicates) {
 
+    public String getEmptyPathReplacement() { return emptyPathReplacement; }
+
+    public void setEmptyPathReplacement(final String emptyPathReplacement) { this.emptyPathReplacement = emptyPathReplacement; }
```
Code style bit missed here

vlascik replied:
pushed a new commit to fix

crazyproger commented on src/ru/crazycoder/plugins/tabdir/TitleFormatter.java:
```diff
@@ -109,6 +105,9 @@ private static String removeDuplicates(final String key, Set<String> neighbours)
     public static String format(List<String> prefixes, String tabName, FolderConfiguration configuration) {
         String joinedPrefixes = joinPrefixes(prefixes, configuration);
+        if (joinedPrefixes.isBlank()) {
+            return configuration.getEmptyPathReplacement() + MessageFormat.format(configuration.getTitleFormat(), joinedPrefixes, tabName);
```
MessageFormat.format calculated in both cases, can be extracted to variable

vlascik replied:
ok, moved it to a variable in a new commit

crazyproger commented on src/ru/crazycoder/plugins/tabdir/SameFilenameTitleProvider.java:
```diff
@@ -110,28 +107,36 @@ private String titleRelativeTo(final VirtualFile file, final FolderConfiguration
+        if (emptyPrefix.isBlank()) {
+            emptyPrefix = globalEmptyPrefix;
         }
+        return emptyPrefix + file.getPresentableName();
```
Could you please describe(or point me) what is the purpose of `emptyPrefix`?
(to be honest, if it unit test can be provided - this is the best variant)

vlascik replied:
So imagine your project has a structure like this
```
my-project/tests/test1.js
my-project/package.json
```

If you configure tabdir for a directory `my-project/`  with always show path relative to `my-project/` and for a directory `my-project/tests` with always show relative to `my-project/tests` and table title format `[{0}] {1}`, you will end up with an empty path (`[{0}]` part) for both `my-project/tests/test1.js` and `my-project/package.json` and you're not going to know which came from where - tab titles are only going to show `test1.js` and `package.json`.

If you configure empty file prefix for `my-project` as e.g. `[MP] ` and empty prefix for `my-project/tests` as e.g. `[T] `, you will end up with tab titles `[MP] package.json` and `[T] test1.js` instead, and you'll know which file came from which directory.

There are probably better examples for this, but I hope it makes sense.

crazyproger replied:
Got it. Case you've described is very similar to Scopes+File Colors builtin feature(but instead of prefix you can get custom tab colour for different folders).
It might be rare case, but since mostly all Tabdir users - are rare auditory, lets add this feature.

But I think this config in Global settings is unnecessary, can not imagine case where user want add some prefix to _all_ tabs in project. It should be available only in per-folder configuration. If you agree - new issue should be filled for future release to fix this.

crazyproger reviewed (commented):
Thank you for this PR. 
I've asked one question about Empty Path Replacement feature. 
And pointed to two minor CS problems.

Before merging I want to understand the purpose of Empty Path feature, after this I will immediately submit update to the JB repository.

vlascik commented:
> Before merging I want to understand the purpose of Empty Path feature, after this I will immediately submit update to the JB repository.

There are still some exception(s?) thrown by this addon, people have reported on it in several Issues, and I'm not sure what's going on there. It seems to be about logging control flow exception and/or about reading filesystem without wrapping the calls in IDEA's filesystem call wrapper, something along the lines of `java.lang.Throwable: Read access is allowed from inside read-action (or EDT) only (see com.intellij.openapi.application.Application.runReadAction())`.

I'm trying to track it down, but I haven't been doing Java for actual decades :D so if you can look at it too, that would be great.

I'm also trying to add one more feature, regexp replacements for tab titles, but that will be a separate PR later.

crazyproger commented:
Sadly, but by some reason I'm not able to update plugin in JB repository any more(
Plugin should be submitted to JB as new one, from different account.

@vlascik If you want - feel free to fork this repo and maintain it further. I will add redirect in this repo to your(or another) repository.

vlascik commented:
> Sadly, but by some reason I'm not able to update plugin in JB repository any more( Plugin should be submitted to JB as new one, from different account.
> 
> @vlascik If you want - feel free to fork this repo and maintain it further. I will add redirect in this repo to your(or another) repository.

I just submitted it as a new plugin at https://plugins.jetbrains.com/plugin/24528-tabdir-redux , should be approved in 2 days. The repo is at https://github.com/vlascik/Tabdir.

Seems that I fixed the exceptions people were complaining about in issues.

I also added an option to use Regex replacements on tab titles.

Users will have to uninstall your version before installing a new one.
